repository.c

```diff
@@ -46,7 +46,7 @@ static void repo_set_commondir(struct repository *repo,
 {
 	struct strbuf sb = STRBUF_INIT;
 
-	free(repo->commondir);
+	FREE_AND_NULL(repo->commondir);
```

Comment:
Wait, didn't you say that the commit that introduced the problem was not yet upstreamed? But I see https://github.com/gitgitgadget/git/pull/1402/commits/b0f1a413cd902b6c8ad9ed935ac58e67f5206c76 being the identical patch, but targeting upstream's default branch...

Reply:
Yes, actually I first thought that it will be applicable there, but then started to investigate further and came to conclusion that it quite probably would not be relevant there and closed my PR.

Anyway, I've seen that you hinted in that PR something like that it still can be important in upstream, so I'll read your comments more thoroughly and then decide whether to continue working with upstream instead.

Reply:
The bug described is not present in upstream (at least in the exact same scenario).

Reply:
I am actually rather convinced that [the upstream contribution](https://github.com/gitgitgadget/git/pull/1402) is the way we should go forward with: the patch is clearly fixing a bug that is caused by the combination of two upstream patches, and even if the current symptom cannot be seen with upstream's version, it is still a bug in upstream.

Reply:
Could you elaborate, please? I'm not aware of implicit assumptions taken in the codebase, and for me it's quite likely that they are implying that the code path in question should not request for config values of the global repository object... If that's not the case, I would be happy to make an upstream contribution with corresponding reasoning.

Reply:
> Could you elaborate, please?

In https://github.com/gitgitgadget/git/commit/f9b7573f6b0039d298de826e22c636db79b9c919#diff-8107da631aae41d1abfe44a6e2ee403d79b2bdf07398eb1f7737fefeb073c0acR43, code was added that releases the memory that `repo->commondir` points to, but it does not set that pointer to `NULL` because the very next line sets it to something different.

However, in https://github.com/gitgitgadget/git/commit/357a03ebe9e04329705a46ff36d526c5ab0c3ebc#diff-8107da631aae41d1abfe44a6e2ee403d79b2bdf07398eb1f7737fefeb073c0acR59-R66, quite a bit of code (and a call to another function) was introduced between releasing that memory and setting it again.

This introduced the fragility in the code that eventually lead to the code path (currently only a problem in Git for Windows' fork).

One might argue that the bug is in the `windows.appendAtomically` code, but this would be an incorrect assessment: the fragile code was introduced by 357a03ebe9e04329705a46ff36d526c5ab0c3ebc and it was only a matter of time until a patch would trigger this fragility, in this instance 61cbb768c1f8858fa0a276262a6c6dd1736cba81.

What 61cbb768c1f8858fa0a276262a6c6dd1736cba81 does is totally legitimate: it asks whether there is a `repo->commondir` to work with, and if there is, uses it. It's the fault of 357a03ebe9e04329705a46ff36d526c5ab0c3ebc  that there is an unusable, non-`NULL` `repo->commondir`.

Reply:
Thanks, exactly what I thought (actually). I'll prepare the PR to upstream and get this discussion linked in it.

Reply:
> I'll prepare the PR to upstream

@zwerdna you could simply reopen https://github.com/gitgitgadget/git/pull/1402 :-)